### PR TITLE
[RLlib] Fix test backward compatibility test for RL Modules

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

The backward compatibility test includes the `explore` parameter, which we don't with RLModule's enabled. 
The test will soon be deprecated, since we are deprecating legacy config dicts.

This PR makes it so that one can still use the legacy config dict syntax, but not with the legacy `explore` option, which is orthogonal to the legacy config dicts.

### CODE REVIEW

**Review by kouroshHakha — COMMENTED** [review]
I think this needs a little bit more thinking. I think having a flag that chooses between forward_inference() and forward_exploration() during sampling still makes sense. So if you look at `torch_policy_v2._compute_action_helper()` you'll see that the functionality is already implemented so that if explore = None or False, forward_inference() is used, if explore=True, forward_exploration() is used. Given this context, I think the validation function is actually wrong (I wrote that initially but I think I was wrong back then about the logic). We want to error out on custom exploration behavior when RLModule is enabled since it's not how users can customize exploration behavior anymore. So I think the right fix is to change the validation error checking logic to check if RLModule is enabled AND custom exploration is set instead of the flag.

**Comment by ArturNiederfahrenhorst** [comment]
@kouroshHakha I tried erroring out when setting exploration_config first.
The issue with that is that the Algorithm base class has an exploration config and we merge with it in many places.
Changes this will be very time consuming and don't improve experience much.
I think a deprecation warning is also enough here - we can simply warn users that setting an exploration config has no effect.

**Comment by ArturNiederfahrenhorst** [comment]
@kouroshHakha I went back and forth on this PR because the changes have pretty wide-ranging consequences.
I'm lay out the problem:

- We have two conflicting configuration options `_enable_rl_module_api` and `exploration_config`
- We can not know which one we set first, hence we can not error out in both cases.
     -> If we error out on both and both are True/{...} by default (instead of False/{}), setting either will trigger the error.
- A check in `AlgorithmConfig.validation()` also catches the cases that we test when we set _enable_rl_modules by default.
    -> This is problematic because we either have to set the `exploration_config={}` by default for PPO when settting `_enable_rl_modules=True` by default there, which brings up problems for all test cases where we set _enable_rl_modules=False OR we circumvent the validation, which is what I've done in this PR with a new paramter `_validate_exploration_conf_and_rl_modules`.

Setting both is an uncommon scenario. 
So I propose that we only error out on AlgorithmConfig.validate() if both are set.
What this means is that instead of only calling `config.rl_modules(_enable_rl_module_api=True)`, one also has to set `config._exploration_config={}`.

File: rllib/algorithms/algorithm_config.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm_config.py:423]: This option makes it possible to disable the Error on "_enable_rl_module_api" and "exploration_config!={}".
This is useful because we need to circumvent this temporarily when switching to `_enable_rl_module_api=True` in the PPOConfig, but still doing `_enable_rl_module_api=False` in some other places where these test would be red otherwise.

File: rllib/algorithms/crr/crr.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/crr/crr.py:53]: Not critical, just changing this to active voice, which should be the default.

File: rllib/algorithms/dreamer/dreamer.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/dreamer/dreamer.py:104]: This was a typo from some time ago

File: rllib/policy/eager_tf_policy.py
Comment (ArturNiederfahrenhorst) [inline rllib/policy/eager_tf_policy.py:587]: self.exploration becomes None if exploration_config is {}.

File: rllib/algorithms/appo/tests/tf/test_appo_learner.py
Comment (kouroshHakha) [inline rllib/algorithms/appo/tests/tf/test_appo_learner.py:70]: Add `TODO (Artur): deprecate exploration_config for APPO when RLModuleAPI is used. `

File: rllib/algorithms/algorithm_config.py
Comment (kouroshHakha) [inline rllib/algorithms/algorithm_config.py:335]: TODO (Artur): Test which algorithms need `exploration_config`'s default old value and which don't on our ci. Then deprecate their default behavior so that for all the derived algorithms the default becomes empty dict.

File: rllib/algorithms/dreamer/dreamer.py
Comment (kouroshHakha) [inline rllib/algorithms/dreamer/dreamer.py:125]: is there a difference between empty dict or None here? For consistency I'd like to use `{}` if possible.

File: rllib/algorithms/dt/dt.py
Comment (kouroshHakha) [inline rllib/algorithms/dt/dt.py:81]: This is offline RL, it should not need `self.exploration_config`'s default of stochasting sampling.

File: rllib/algorithms/crr/crr.py
Comment (kouroshHakha) [inline rllib/algorithms/crr/crr.py:64]: This is offline RL, it should not need self.exploration_config's default of stochasting sampling.

File: rllib/algorithms/impala/tests/test_impala_learner.py
Comment (kouroshHakha) [inline rllib/algorithms/impala/tests/test_impala_learner.py:79]: TODO (Artur): See if setting Impala's default to {} pose any issues. Deprecate the current default and set it to {}.

Comment (kouroshHakha) [reply rllib/algorithms/impala/tests/test_impala_learner.py:79]: Use better wording :)

File: rllib/algorithms/marwil/marwil.py
Comment (kouroshHakha) [inline rllib/algorithms/marwil/marwil.py:97]: offline RL

File: rllib/policy/eager_tf_policy_v2.py
Comment (kouroshHakha) [inline rllib/policy/eager_tf_policy_v2.py:120]: default should be False.

File: rllib/policy/eager_tf_policy_v2.py
Comment (kouroshHakha) [inline rllib/policy/eager_tf_policy_v2.py:477]: is `if self.exploration:` something you can do instead of `if getattr(self, "exploration", None)`? doesn't `self.exploration` always exist? you set it either to None or something else in your constructor.

File: rllib/policy/eager_tf_policy_v2.py
Comment (kouroshHakha) [inline rllib/policy/eager_tf_policy_v2.py:732]: default it to False.

Comment (kouroshHakha) [reply rllib/policy/eager_tf_policy_v2.py:732]: `if not self.config.get("_enable_rl_module_api", False) and self.exploration` ?

File: rllib/policy/policy.py
Comment (kouroshHakha) [inline rllib/policy/policy.py:1445]: same, `if self.exploration is not None`?

File: rllib/policy/torch_policy_v2.py
Comment (kouroshHakha) [inline rllib/policy/torch_policy_v2.py:180]: same comments I left in the previous policy file

**Review by kouroshHakha — COMMENTED** [review]
Some of the algorithms don't actually use `self.exploration_config` by default anyways. So we should just later check which algorithms don't need and remove the Stochastic Sampling exploration term. In this PR we can do it for offline RL algorithms. For online one, we maybe using them. 

Also I am confused on when `self.exploration` is `{}` vs. `None` it's important to pick only one to indicate no exploration.

**Comment by kouroshHakha** [comment]
For approval please show 1 snapshot of the target test passing after RLModule is enabled. Then after disabling all the normal tests should still pass.

File: rllib/algorithms/dreamer/dreamer.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/dreamer/dreamer.py:125]: Altough it won't make a difference in tests, we have some code that assumes that exploration config is always a dict, so you are right.

File: rllib/algorithms/crr/crr.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/crr/crr.py:67]: Removing it does not work. I've left a TODO.

File: rllib/algorithms/marwil/marwil.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/marwil/marwil.py:100]: Removing it does not work. I've left a TODO.

**Comment by ArturNiederfahrenhorst** [comment]
Screenshot 2023-04-02 at 08 04 53

File: rllib/algorithms/algorithm_config.py
Comment (kouroshHakha) [inline rllib/algorithms/algorithm_config.py:1010]: `config.exploration_config = {}` There should be no underscore.

**Review by kouroshHakha — COMMENTED** [review]
Let's get the remaining of tests pass and we should be good to go. While we can go into those tests and turn `exploration_config` to `{}`, I wanna know if it's possible to set the default exploration config to {} in the algorithms that support RLModules today. and fix those test that fail to use `stochastic sampling`. How about that? Then if users want to use RLModules don't need to explicitly set `exploration_config` which is extra cognitive load.

File: rllib/algorithms/algorithm_config.py
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/algorithm_config.py]: Thanks!

**Comment by ArturNiederfahrenhorst** [comment]
Link to build before disable ppo by default again: https://buildkite.com/ray-project/oss-ci-build-pr/builds/17316#018748ff-b5c7-4411-8930-017d1472926e

Screenshot 2023-04-03 at 14 51 44

File: rllib/algorithms/algorithm_config.py
Comment (kouroshHakha) [inline rllib/algorithms/algorithm_config.py:2422]: ok yes. This is the fix.

**Review by kouroshHakha — APPROVED** [review]
ok let's see if this passes.

**Comment by ArturNiederfahrenhorst** [comment]
Screenshot 2023-04-03 at 20 50 30

**Review by kouroshHakha — APPROVED** [review]

**Comment by kouroshHakha** [comment]
@zhe-thoughts Needs your approval.

**Comment by zhe-thoughts** [comment]
@sven1977 could you take a look first? Thanks (A committer should review anyways)

File: rllib/algorithms/algorithm_config.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm_config.py:337]: @sven1977 The idea here is that AlgorithmConfig does not include exploration config so that we don't end up with that being the default everywhere anymore.
This is important because we sometimes merge configs with AlgorithmConfig to get sensible defaults for values.

File: rllib/algorithms/algorithm_config.py
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/algorithm_config.py:861]: Makes it so that we set exploration_config to {} instead of only setting enable_rl_module_api.

**Comment by ArturNiederfahrenhorst** [comment]
@sven1977 The latest commit is cosmetic only, just changed a warning. So this should not change anything regarding Kourosh's approval.

File: rllib/algorithms/algorithm_config.py
Comment (sven1977) [reply rllib/algorithms/algorithm_config.py:337]: Makes sense. I approved. Waiting for @zhe-thoughts for approval so we can merge after the master freeze.

File: rllib/tests/backward_compat/test_backward_compat.py
Comment (sven1977) [inline rllib/tests/backward_compat/test_backward_compat.py:132]: Nice catch!

File: rllib/policy/torch_policy_v2.py
Comment (sven1977) [inline rllib/policy/torch_policy_v2.py:460]: This might break for algos not using the standard `Algorithm` path, like ES, ARS, .... But tests seem to be ok.

**Review by sven1977 — APPROVED** [review]
Looks ok to me. Approved. Thanks for the fixes @ArturNiederfahrenhorst !

**Review by zhe-thoughts — APPROVED** [review]

**Comment by zhe-thoughts** [comment]
@ArturNiederfahrenhorst please also work with @clarng to make sure this is included in the release branch (we are too close to cutting it now)

**Pull request merged**
